src/target/cortexar.c

```diff
@@ -1074,9 +1093,16 @@ static void cortexar_mem_handle_fault(target_s *const target, const char *const
  * This reads memory by jumping from the debug unit bus to the system bus.
  * NB: This requires the core to be halted! Uses instruction launches on
  * the core and requires we're in debug mode to work. Trashes r0.
+ * If core is not halted, falls back to adiv5 mem read.
  */
 static void cortexar_mem_read(target_s *const target, void *const dest, const target_addr64_t src, const size_t len)
 {
+	/* If system is not halted, perform basic adiv5 memory read. */
+	if (!(cortex_dbg_read32(target, CORTEXAR_DBG_DSCR) & CORTEXAR_DBG_DSCR_HALTED)) {
+		adiv5_mem_read(cortex_ap(target), dest, src, len);
```

Comment:
Fair enough. I was hoping to avoid the use of in-function halt/resume but I guess it's unavoidable in certain scenarios. Will backtrack this change to my previous revision.